### DISCUSSION TITLE
Why is my process hanging on exit?

### DISCUSSION THREAD

**Discussion by eliphatfs** [seaweedfs/seaweedfs#5380] [Q&A] [answered]
Hi; my setup is 1 master + 1 filer + 8 volume servers. I use `weed mount` to mount the fs on host nodes and run docker containers that mounts the directory with `-v` options on the nodes. The docker program is a ML training code that reads files and runs computation. However, sometimes at process exit (or when `kill`-ed), the docker program would hang. The process stops after restarting `weed mount` but that also interrupts other services relying on the storage on the same node.

The docker program is in uninterruptible sleep (D) state and `kill -9` cannot kill it.
Using `/proc/<pid>/stack` I was able to trace down to `flip_flush` (caused by `closefd`) in linux kernel https://github.com/torvalds/linux/blob/master/fs/open.c#L1505 and `request_wait_answer` in `fuse_flush`. However, gdb and `pstack` hangs until program exit without any output and `kill -QUIT` does not provide an output either. I guess some other system signals come earlier and prevented these attach/dump from working. So I don't know what is going on the golang side. I didn't notice any possibility for the `flush` to hang forever on the golang side.

Any assistance would be appreciated. Thank you.

**System Setup**
- `/usr/local/bin/weed server -volume=0 -filer -dir=/weedfs`
- `/usr/local/bin/weed volume -max=400 -dir=/weedfs -mserver=10.8.149.13:9333` on 8 machines different from the master
- These commands are done by systemctl services with setups exactly as the wiki page.
- OS version: Ubuntu 22.04
- output of `weed version`: version 8000GB 3.63 54d7748a4a54d94a31ce04d05db801faeff4f690 linux amd64
- if using filer, show the content of `filer.toml`: no `filer.toml`
- security:
```
[access]
ui = false

[grpc]
ca = "/etc/ariesdockerd/certs/Aries_SeaweedFS_CA.crt"

[grpc.volume]
cert = "/etc/ariesdockerd/certs/volume01.crt"
key  = "/etc/ariesdockerd/certs/volume01.key"

[grpc.master]
cert = "/etc/ariesdockerd/certs/master01.crt"
key  = "/etc/ariesdockerd/certs/master01.key"

[grpc.filer]
cert = "/etc/ariesdockerd/certs/filer01.crt"
key  = "/etc/ariesdockerd/certs/filer01.key"

[grpc.client]
cert = "/etc/ariesdockerd/certs/client01.crt"
key  = "/etc/ariesdockerd/certs/client01.key"
```

**Comment by NyaMisty** [comment]
The hang is because program has performed read or write operations on the weed-mounted fs, and the pending operation (including read/write/close/flush/list/..., actually all FS operations) must be finished before shutdown.

When you use gdb/pstack or any ptrace-based tools, they won't be able to attach the process because the program already entered kernel space (because FS operations) and they must wait until the operation was finished by `weed mount`. Terminating `weed mount` directly interrupted all FS operations, thus the program exited directly without any wait.

**Reply by eliphatfs** [reply]
Thank you very much for the explanation.

why the fs does not interrupt all operations at shutdown itself… it is an obvious source of deadlock with no solution because i cannot know what logic (maybe in the library, or as a cluster manager, from users) is executing and using the storage mount. Remounting interrupts other users.

**Comment by Dando-Real-ITA** [comment, marked as answer]
I am having the exact same problem when terminating bash scripts that are hosted on seaweedfs and mounted with weed mount.

I found here: https://netflixtechblog.com/debugging-a-fuse-deadlock-in-the-linux-kernel-c75cd7989b6d

That the solution would be for the fuse implementation ( go-fuse? ) to not wait for flush response on exit

**Reply by eliphatfs** [reply]
The problem seems to be inside the linux kernel; finishing fast on go side can reduce but not fully solve the problem?

**Reply by eliphatfs** [reply]
Thank you very much for your comments. I have a good vision on the problem now.

**Reply by eliphatfs** [reply]
I updated linux kernel to 6.5 but the problem still occurs...

**Reply by eliphatfs** [reply]
It does become an interruptible sleep in 'ps', but cannot be interrupted anyway... So stupid I would think.

**Reply by Dando-Real-ITA** [reply]
In my case I improved ( and seems removed ) the problem by using "sync" at all exit points, also I had a while(true) watchdog in the main process, that I moved to a background function.
Maybe the docker program has something similar?
